Sources/Brave/Frontend/Browser/Playlist/Controllers/PlaylistCarplayController.swift

```diff
@@ -811,210 +772,91 @@ extension PlaylistCarplayController {
-          }
-
-          // Attempt to play the stream
-          if let url = URL(string: item.src) {
-            self.load(url: url, autoPlayEnabled: true)
-              .handleEvents(receiveCancel: {
-                if !isPlaying {
-                  PlaylistMediaStreamer.clearNowPlayingInfo()
-                }
-                completion?(.cancelled)
-              })
-              .sink(
-                receiveCompletion: { status in
-                  switch status {
-                  case .failure(let error):
-                    if !isPlaying {
-                      PlaylistMediaStreamer.clearNowPlayingInfo()
-                    }
-
-                    switch error as? MediaPlaybackError {
-                    case .cancelled:
-                      if !isPlaying {
-                        PlaylistMediaStreamer.clearNowPlayingInfo()
-                      }
-                      completion?(.cancelled)
-                    case .other(let err):
-                      completion?(.other(err))
-                    default:
-                      completion?(.other(error))
-                    }
-                  case .finished:
-                    break
-                  }
-                },
-                receiveValue: { [weak self] _ in
-                  guard let self = self else {
-                    if !isPlaying {
-                      PlaylistMediaStreamer.clearNowPlayingInfo()
-                    }
-                    completion?(.cancelled)
-                    return
-                  }
-
-                  PlaylistMediaStreamer.setNowPlayingInfo(item, withPlayer: self.player)
-                  completion?(.none)
-                }
-              ).store(in: &self.assetLoadingStateObservers)
-            Logger.module.debug("Playing Live Video: \(self.player.isLiveMedia)")
-          } else {
-            if !isPlaying {
-              PlaylistMediaStreamer.clearNowPlayingInfo()
-            }
-            completion?(.expired)
-          }
-        }
-      ).store(in: &assetStateObservers)
+    var item = item
+    
+    do {
+      item = try await mediaStreamer.loadMediaStreamingAsset(item)
+      if !isPlaying { PlaylistMediaStreamer.clearNowPlayingInfo() }
+    } catch {
+      if !isPlaying { PlaylistMediaStreamer.clearNowPlayingInfo() }
+      throw error
+    }
+    
+    // Item can be streamed
+    guard let url = URL(string: item.src)
+    else {
+      if !isPlaying { PlaylistMediaStreamer.clearNowPlayingInfo() }
+      throw PlaylistMediaStreamer.PlaybackError.expired
+    }
+    
+    // Attempt to play the stream
+    do {
+      try await load(url: url, autoPlayEnabled: true)
+      if !isPlaying { PlaylistMediaStreamer.clearNowPlayingInfo() }
+      PlaylistMediaStreamer.setNowPlayingInfo(item, withPlayer: self.player)
+      Logger.module.debug("Playing Live Video: \(self.player.isLiveMedia)")
+    } catch {
+      if !isPlaying { PlaylistMediaStreamer.clearNowPlayingInfo() }
```

Comment:
Seems like what you want here is a `defer` at the top of this method so that it executes when its returning right?

```swift
defer { 
  if !isPlaying { PlaylistMediaStreamer.clearNowPlayingInfo() 
}
```

Reply:
Though not sure if you need to do it prior to that `PlaylistMediaStreamer.setNowPlayingInfo(item, withPlayer: self.player)` call and then not at all if it succeeds…

Reply:
When it succeeds, I don't want it called. I only want it called when it is `not` playing.